File: src/routes/labs.py
import json
import yaml
import os
import flask

import conf

app = flask.Blueprint('labs', __name__)


@app.route('', methods=['GET'])
@app.route('/', methods=['GET'])
def help():
    return """
Usage: curl <host>/labs/<api> -H 'Content-Type: application/json'

GET:
    getdata                     Get all of labs info.
                                return: list
    <labId>/getdata             Get lab info which you select.
                                return: dict
    <labId>/file/<path>         Download the file that is documented in the frontendvariable in config.yaml from <labId> dir.
                                return: file
    <labId>/file/description    Download the description file of <labId>.
                                return: file
"""

@app.route('/getdata',methods=['GET'])
def getLabs():
    if conf.stoping:
        return json.dumps([])
    result = []
    for a in sorted(os.listdir('judge/labs'), reverse=True):
        data = json.loads(getLab(a))
        if data is not None:
            result.append(data)

    return json.dumps(result)

@app.route('/<string:labId>/getdata',methods=['GET'])
def getLab(labId):
    if conf.stoping:
        return json.dumps(None)
    labId = labId.split('/')[0]
    result = None
    if os.path.isdir(f'judge/labs/{labId}') and os.path.isfile(f'judge/labs/{labId}/config.yaml'):
        with open(f'judge/labs/{labId}/config.yaml', 'r') as f:
            try:
                labconfig = yaml.load(f, Loader=yaml.FullLoader)
            except yaml.scanner.ScannerError:
                labconfig = {}
        if 'promissions' not in labconfig:
            labconfig['promissions'] = []
        if 'frontendvariable' not in labconfig:
            labconfig['frontendvariable'] = []
        if 'deadlines' not in labconfig:
            labconfig['deadlines'] = []
        if 'checkpoints' not in labconfig:
            labconfig['checkpoints'] = {}
        result = {'id':labId, 'contents':labconfig['frontendvariable'], 'checkpoints':labconfig['checkpoints'], 'promissions':labconfig['promissions'], 'deadlines':labconfig['deadlines']}
    return json.dumps(result)

@app.route('/<string:labId>/file/<path:path>',methods=['GET'])
def download(labId, path):
    if conf.stoping:
        return ""
    if os.path.isfile(f'judge/labs/{labId}/download/{path}'):
        return flask.send_from_directory(f'judge/labs/{labId}/download/', path, as_attachment=True)
    else:
        return_result = {'code': 404, 'Success': False,
                         "Message": "The website is not available currently"}
        return flask.jsonify(return_result), 404

@app.route('/<string:labId>/file/description',methods=['GET'])
def description(labId):
    if conf.stoping:
        return ""
    with open(f'judge/labs/{labId}/config.yaml', 'r') as f:
        labdata = yaml.load(f, Loader=yaml.FullLoader)
    
    if os.path.isfile(f'judge/labs/{labId}/{labdata["description"]}'):
        return flask.send_from_directory(f'judge/labs/{labId}/', labdata['description'], as_attachment=True)
    else:
        return_result = {'code': 404, 'Success': False,
                         "Message": "The website is not available currently"}
        return flask.jsonify(return_result), 404

@app.errorhandler(404)
def page_not_found(e):
    return_result = {'code': 404, 'Success': False,
                     "Message": "The website is not available currently"}
    return jsonify(return_result), 404


@app.errorhandler(403)
def forbidden(e):
    return_result = {'code': 403, 'Success': False,
                     "Message": "The website is not available currently"}
    return jsonify(return_result), 403

File: router/labs/labs.go
package labs
import (
    "path"
    "github.com/gin-gonic/gin"
    
    "github.com/NCKU-NASA/nasa-judge-lib/schema/lab"

    "github.com/NCKU-NASA/nasa-judge-api/middlewares/auth"
    "github.com/NCKU-NASA/nasa-judge-api/utils/filepath"
    "github.com/NCKU-NASA/nasa-judge-api/utils/errutil"
)

var router *gin.RouterGroup

func init() {
    err := lab.Commit("all")
    if err != nil {
        panic(err)
    }
}

func Init(r *gin.RouterGroup) {
    router = r
    router.GET("/", auth.CheckIsTrust, help)
    router.GET("/:labId/get", auth.CheckIsTrust, get)
    router.GET("/:labId/commit", auth.CheckIsTrust, commit)
    router.GET("/:labId/file/:path", auth.CheckIsTrust, file)
}

func help(c *gin.Context) {
    c.String(200, `
Usage: curl <host>/labs/<api> -H 'Content-Type: application/json'

<labId> is lab ID. If <labId> is 'all' it mean all labs.

GET:
    <labId>/get                 Get lab info which you select.
                                return: dict
    <labId>/file/<path>         Download the file that is documented in the frontendvariable in config.yaml from <labId> dir.
                                return: file
    <labId>/file/description    Download the description file of <labId>.
                                return: file
`)
}

func commit(c *gin.Context) {
    labId := c.Param("labId")
    err := lab.Commit(labId)
    if err != nil {
        errutil.AbortAndError(c, &errutil.Err{
            Code: 500,
            Msg: err.Error(),
        })
        return
    }
    c.String(200, "Commit success")
}

func get(c *gin.Context) {
    labId := c.Param("labId")
    if labId == "all" {
        labs, err := lab.GetLabs()
        if err != nil {
            errutil.AbortAndError(c, &errutil.Err{
                Code: 500,
                Msg: err.Error(),
            })
            return
        }
        c.JSON(200, labs)
        return
    }
    nowlab, err := lab.GetLab(labId)
    if err != nil {
        errutil.AbortAndStatus(c, 404)
        return
    }
    c.JSON(200, nowlab)
}

func file(c *gin.Context) {
    labId := c.Param("labId")
    file := c.Param("path")

    if labId == "all" {
        errutil.AbortAndStatus(c, 404)
        return
    }
    labdata, err := lab.GetLab(labId)
    if err != nil {
        errutil.AbortAndStatus(c, 404)
        return
    }
    
    nowpath, err := filepath.VerifyPath(path.Join(lab.LabDir, labId, "download", file), path.Join(lab.LabDir, labId, "download"))
    if file == "description" {
        nowpath, err = filepath.VerifyPath(path.Join(lab.LabDir, labId, labdata.Description), path.Join(lab.LabDir, labId))
    }
    if err != nil {
        errutil.AbortAndStatus(c, 404)
        return
    }

    c.FileAttachment(nowpath, path.Base(nowpath))
}
